lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+                std::is_same_v<A, Pa::FlushStmt>) {
+    for (const auto &spec : stmt.v) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::InquireStmt>) {
+    for (const auto &spec : std::get<std::list<Pa::InquireSpec>>(stmt.u)) {
+      if (std::holds_alternative<Pa::ErrLabel>(spec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+template <typename A>
+constexpr bool hasEorLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::EorLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt>) {
+    for (const auto &waitSpec : stmt.v) {
+      if (std::holds_alternative<Pa::EorLabel>(waitSpec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+template <typename A>
+constexpr bool hasEndLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::EndLabel>(control.u)) {
+        return true;
+      }
+    }
+  }
+  if constexpr (std::is_same_v<A, Pa::WaitStmt>) {
+    for (const auto &waitSpec : stmt.v) {
+      if (std::holds_alternative<Pa::EndLabel>(waitSpec.u)) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool hasAltReturns(const Pa::CallStmt &callStmt) {
+  const auto &args{std::get<std::list<Pa::ActualArgSpec>>(callStmt.v.t)};
+  for (const auto &arg : args) {
+    const auto &actual{std::get<Pa::ActualArg>(arg.t)};
+    if (std::holds_alternative<Pa::AltReturnSpec>(actual.u)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Determine if `callStmt` has alternate returns and if so set `e` to be the
+/// origin of a switch-like control flow
+void altRet(AST::Evaluation &e, const Pa::CallStmt *callStmt,
+            AST::Evaluation *cstr) {
+  if (hasAltReturns(*callStmt)) {
+    e.setCFG(AST::CFGAnnotation::Switch, cstr);
+  }
+}
+
+template <typename A>
+void ioLabel(AST::Evaluation &e, const A *s, AST::Evaluation *cstr) {
+  if (hasErrLabel(*s) || hasEorLabel(*s) || hasEndLabel(*s)) {
+    e.setCFG(AST::CFGAnnotation::IoSwitch, cstr);
+  }
+}
+
+void annotateEvalListCFG(std::list<AST::Evaluation> &evals,
+                         AST::Evaluation *cstr) {
+  bool nextIsTarget = false;
+  for (auto &e : evals) {
+    e.isTarget = nextIsTarget;
+    nextIsTarget = false;
+    if (e.isConstruct()) {
+      annotateEvalListCFG(*e.getConstructEvals(), &e);
+      // assume that the entry and exit are both possible branch targets
+      nextIsTarget = true;
+    }
+    if (e.isActionStmt() && e.lab.has_value()) {
+      e.isTarget = true;
+    }
+    std::visit(
+        Co::visitors{
```

Comment:
These items could be combined.

Reply:
Same answer as for other comments about combining code with templates, I agree they could, I think it also makes them easier to reason about right now.

Reply:
Eventually condensed as suggested.